### PR TITLE
[PR] Refactor routing to use Strategy pattern

### LINKED ISSUE #319: [REFACTOR] Refactor Navigation Routing With Strategy Pattern (Driving/Walking/Tunnel/Shuttle)
🛠️ **Refactoring Type**  
Introduce Design Pattern: Strategy Pattern: Refactor orchestration logic into mode-specific route strategies and a route planner context.

🧾 **Reason**  
Current route planning behavior is concentrated in one large hook with many branches and effects, which makes it harder to maintain, extend, and test.

This refactor improves robustness and flexibility by isolating transport-specific behavior (driving, outdoor walking, tunnel walking, shuttle) into independent strategy modules.

🔗 **Related PR/Branch**  
PR: N/A this was brought up separately in code review.
Branch name: refactor/navigation-strategy-pattern

💾 **Commit**  
[Commit Link](https://github.com/SOEN-390-MakeSoft/SOEN-390-MakeSoft/commit/a0c3620cb27774661e46baac7fb1a9fc9267f1ce)

📂 **Affected Area**

**Primary hook and route orchestration:**
- useNavigationBetweenBuildings.ts

**Integration surface:**
- MapScreen.tsx

**Behavior/tests to preserve:**
- useNavigationBetweenBuildings-Test.ts

📌 **Resolution Note**  
When this refactoring task is completed and committed, please comment the commit hash/number on this issue itself.

**Definition of done:**

- RouteStrategy contract and strategy implementations exist for driving, outdoor walking, tunnel walking, and shuttle.
- Hook delegates route computation to strategy layer instead of in-effect mode branching.
- Existing route behavior and tests remain green.
- New unit tests cover each strategy independently.

### PR DESCRIPTION
**Pull request by MassPaol**
## Related Work
- Closes #319
- Issue link: https://github.com/DaniK1001/SOEN-390-MakeSoft/issues/319
- Commit: a0c3620cb27774661e46baac7fb1a9fc9267f1ce
- Commit link: https://github.com/DaniK1001/SOEN-390-MakeSoft/commit/a0c3620cb27774661e46baac7fb1a9fc9267f1ce

## Summary
This PR refactors routing logic in the mobile navigation flow to use the Strategy pattern.

The previous approach had route-mode logic concentrated in one hook. This PR introduces a planner and mode-specific strategies so each route mode is isolated, easier to test, and safer to evolve.

## What Changed
- Added strategy contracts and shared route domain types.
- Added concrete strategies for:
  - Driving route
  - Outdoor walking route
  - Tunnel walking route
  - Shuttle route
- Added a RoutePlanner orchestration layer.
- Refactored useNavigationBetweenBuildings to delegate route computations to RoutePlanner and strategies.
- Preserved existing behavior for transport mode selection and route UI state.

## Files Changed
- mobile-app/hooks/useNavigationBetweenBuildings.ts
- mobile-app/services/navigation/RoutePlanner.ts
- mobile-app/services/navigation/types.ts
- mobile-app/services/navigation/routeUtils.ts
- mobile-app/services/navigation/strategies/ModeRouteStrategy.ts
- mobile-app/services/navigation/strategies/DrivingRouteStrategy.ts
- mobile-app/services/navigation/strategies/OutdoorWalkingRouteStrategy.ts
- mobile-app/services/navigation/strategies/TunnelWalkingRouteStrategy.ts
- mobile-app/services/navigation/strategies/ShuttleStrategy.ts
- mobile-app/services/navigation/strategies/ShuttleRouteStrategy.ts
- mobile-app/__tests__/RoutePlanner.strategy.test.ts
- mobile-app/__tests__/GoogleDirectionsStrategies.test.ts
- mobile-app/__tests__/ShuttleRouteStrategy.test.ts

## Design Pattern Applied
- Pattern: Strategy
- Context: RoutePlanner coordinates interchangeable route strategies.
- Benefit:
  - Better compartmentalization of mode-specific logic
  - Stronger unit-test coverage at strategy layer
  - Lower risk of regressions when adding or modifying route modes

### Added tests
- mobile-app/__tests__/RoutePlanner.strategy.test.ts
- mobile-app/__tests__/GoogleDirectionsStrategies.test.ts
- mobile-app/__tests__/ShuttleRouteStrategy.test.ts

### Full suite run
- Command: npm test -- --runInBand
- Result: 49 passed test suites, 636 passed tests, 0 failed

## Risk and Mitigation
- Risk: route-mode behavior drift after moving logic out of the hook.
- Mitigation:
  - Preserved existing hook-level behavior and state updates.
  - Added focused strategy tests.
  - Verified with full test suite.

## Checklist
- [x] Strategy pattern implemented for navigation routing.
- [x] Existing behavior preserved.
- [x] New unit tests added for planner/strategies.
- [x] Full test suite passing.
- [x] Commit references issue #319.

## PR Notes for Reviewers
Please focus review on:
- Correct parity of route selection behavior in useNavigationBetweenBuildings.
- Strategy boundaries and testability.
- Shuttle strategy parity with previous timing and catchable-departure logic.

### CODE REVIEW

**Comment by MassPaol** [comment]
> hey @MassPaol,
> 
> Thanks for your work on this, this is definitely a great next step forward
> 
> Just a quick thought: we currently have a 2 active PRs, plus another 2 usability testing tasks that should be merged soon.
> 
> Do you think this might introduce any conflicts later on? And more generally, do you see the design patterns here integrating smoothly with those upcoming changes?

Hey @DaniK1001, I don't think there should be any issues, after taking a look. I believe the integration should go smoothly, Just to be safe however, I can wait until those other PRs are merged and review it before merging.

File: mobile-app/__tests__/ShuttleRouteStrategy.test.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    routes: [
+      {
+        overview_polyline: { points: ENCODED_POLYLINE },
+        legs: [{ duration: { value: durationSec } }],
+      },
+    ],
+  };
+}
+
+describe('DefaultShuttleRouteStrategy', () => {
```
Comment (RyanCheung03) [inline mobile-app/__tests__/ShuttleRouteStrategy.test.ts]: Minor: Right now these only validate successful directions responses. Since fetchShuttleSegment() catches failures and returns empty zeroed data, we should also have a test covering a directions fetch failure so it’s clear whether the strategy is expected to return null, partial shuttle info, or suppress directions. Right now that behavior is implicit and untested.`

**Comment by MassPaol** [comment]
Ok perfect, merging.
